packages/utils/src/Keypairs.ts

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (c) 2018-2022, BOTLabs GmbH.
+ *
+ * This source code is licensed under the BSD 4-Clause "Original" license
+ * found in the LICENSE file in the root directory of this source tree.
+ */
+
+import { Keyring } from '@polkadot/keyring'
+import { decodePair } from '@polkadot/keyring/pair/decode'
+import { hexToU8a, isHex } from '@polkadot/util'
+import {
+  ed25519PairFromSeed,
+  randomAsHex,
+  secp256k1PairFromSeed,
+  sr25519PairFromSeed,
+  naclBoxPairFromSecret,
+  keyFromPath,
+  keyExtractPath,
+  mnemonicToMiniSecret,
+} from '@polkadot/util-crypto'
+import { Keypair } from '@polkadot/util-crypto/types'
+import { HexString } from '@polkadot/util/types'
+
+import {
+  VerificationKeyType,
+  KiltKeyringPair,
+  EncryptionKeyType,
+  verificationKeyTypes,
+} from '@kiltprotocol/types'
+
+import { UnsupportedKeyError } from './SDKErrors.js'
+
+export { mnemonicGenerate as generateMnemonic } from '@polkadot/util-crypto'
+
+/**
+ * Randomly generates a new 32 byte secret.
+ *
+ * @returns Secret encoded as a 0x-prefixed hex string.
+ */
+export function generateSecret(): HexString {
+  return randomAsHex(32)
+}
+
+const PairFromSeed: Record<string, (seed: Uint8Array) => Keypair> = {
+  ecdsa: (seed) => secp256k1PairFromSeed(seed),
+  ed25519: (seed) => ed25519PairFromSeed(seed),
+  sr25519: (seed) => sr25519PairFromSeed(seed),
+  x25519: (seed) => naclBoxPairFromSecret(seed),
+}
+
+/**
+ * Creates an encryption or verification key pair from a mnemonic or hex-encoded secret.
+ *
+ * @param type {Verification|Encryption} Key type.
+ * @param mnemonicOrSeed A BIP39 mnemonic or 32 byte hex string.
+ * @param derivationPath Key derivation path beginning with `//` or `/', e.g. `//hard/soft`.
+ * @returns An object containing public and secret keys.
+ */
+export function createKeypair(
+  type: VerificationKeyType | EncryptionKeyType,
+  mnemonicOrSeed: string | HexString,
+  derivationPath: string
+): Keypair {
+  if (!Object.prototype.hasOwnProperty.call(PairFromSeed, type)) {
+    throw new UnsupportedKeyError(type)
+  }
+  const seed = isHex(mnemonicOrSeed)
+    ? hexToU8a(mnemonicOrSeed)
+    : mnemonicToMiniSecret(mnemonicOrSeed)
+  let pair = PairFromSeed[type](seed)
+  if (derivationPath) {
+    const { path } = keyExtractPath(derivationPath)
```

Comment:
I think the derivation path should be applied to the seed

Reply:
sorry, I think I misread the code. It `is` applied to the keypair, which was generated from seed in the next line.